plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java

```diff
@@ -2711,6 +2748,77 @@ public void finishStatisticsCollection(ConnectorSession session, ConnectorTableH
+                        || addFileEntry.getStats().get().getMinValues().isEmpty()
+                        || addFileEntry.getStats().get().getNullCount().isEmpty())
+                // Statistics returns whole path to file, so we need to create corresponding map key for AddFileEntry.
+                .map(addFileEntry -> Map.entry(Location.of(tableHandle.getLocation()).appendPath(addFileEntry.getPath()).toString(), addFileEntry))
+                .collect(toImmutableMap(Entry::getKey, Entry::getValue));
+
+        if (addFileEntriesWithNoStats.isEmpty()) {
+            return;
+        }
+
+        Map<String, DeltaLakeColumnHandle> columnsHandles = getColumns(tableHandle.getMetadataEntry()).stream()
+                .filter(column -> column.getColumnType() == REGULAR)
+                .map(column -> Map.entry(column.getBaseColumnName(), column))
+                .collect(toImmutableMap(Entry::getKey, Entry::getValue));
+
+        List<AddFileEntry> updatedAddFileEntries = computedStatistics.stream()
+                .map(statistics -> {
+                    // Grouping by `PATH_COLUMN_NAME`.
+                    String filePathFromStatistics = VARCHAR.getSlice(statistics.getGroupingValues().get(0), 0).toStringUtf8();
+                    if (addFileEntriesWithNoStats.containsKey(filePathFromStatistics)) {
+                        return Optional.of(prepareUpdatedAddFileEntry(statistics, addFileEntriesWithNoStats.get(filePathFromStatistics), columnsHandles));
+                    }
+                    return Optional.<AddFileEntry>empty();
+                })
+                .flatMap(Optional::stream)
+                .toList();
+
+        if (updatedAddFileEntries.isEmpty()) {
+            return;
+        }
+        try {
+            long createdTime = Instant.now().toEpochMilli();
+            long readVersion = tableHandle.getReadVersion();
+            long commitVersion = readVersion + 1;
+            TransactionLogWriter transactionLogWriter = transactionLogWriterFactory.newWriter(session, tableHandle.getLocation());
+            transactionLogWriter.appendCommitInfoEntry(getCommitInfoEntry(session, commitVersion, createdTime, OPTIMIZE_OPERATION, readVersion));
+            updatedAddFileEntries.forEach(transactionLogWriter::appendAddFileEntry);
+            transactionLogWriter.flush();
+        }
+        catch (Throwable e) {
+            throw new TrinoException(DELTA_LAKE_BAD_WRITE, "Unable to access file system for: " + tableHandle.getLocation(), e);
+        }
+    }
+
+    private AddFileEntry prepareUpdatedAddFileEntry(ComputedStatistics stats, AddFileEntry addFileEntry, Map<String, DeltaLakeColumnHandle> columnsHandles)
+    {
+        DeltaLakeJsonFileStatistics deltaLakeJsonFileStatistics = DeltaLakeComputedStatistics.toDeltaLakeJsonFileStatistics(stats, columnsHandles);
+        try {
+            return new AddFileEntry(
+                    addFileEntry.getPath(),
+                    addFileEntry.getPartitionValues(),
```

Comment:
This is deprecated and supposed not to be used (`// required for JSON serialization; getCanonicalPartitionValues should be used in code instead.`)

i think it actually should be used here, so let's have a comment

```suggestion
                    addFileEntry.getPartitionValues(), // preserve original case without canonicalization
```